File: #2707.extra-characters-in-a-string.cpp
class Solution {
public:
    int minExtraChar(string s, vector<string>& dictionary) {

        int size = s.size();
        vector<int>dp(size + 1, 0);  
        for (int i = size - 1; i >= 0; i--) {
            dp[i] = dp[i + 1] + 1;
            for (auto it: dictionary) {
                if (s.compare(i, it.size(), it) ==0) {
                    dp[i] = std::min(dp[i], dp[i + it.length()]);
                }
            }
        }
        return dp[0];
    }
};

File: #2707.extra-characters-in-a-string.js
/**
 * @param {string} s
 * @param {string[]} dictionary
 * @return {number}
 */
var minExtraChar = function(s, dictionary) {
    let arr = new Array(s.length + 1).fill(-1);
    arr[s.length] = 0;
    for (let i = s.length - 1; i >= 0; i--) {
        arr[i] = arr[i + 1] + 1;
        for (let w of dictionary) {
            let subStr = s.substring(i, i + w.length);
            if (subStr === w){
                let prev = arr[i + w.length];
                if (prev < arr[i]) {
                    arr[i] = prev;
                }
            }
        }
    }
    return arr[0];
};
